Always build SwiftDriver and use it as default compiler driver

artemcm opened the pull request:
Always build (and use) the new [SwiftDriver](https://github.com/apple/swift-driver) as the default compiler driver.
This will make sure that compiler developers are using the new driver when they build the compiler locally and use it.

- Adds a new `build-script` product category: `before_build_script_impl` for products we wish to build before the `impl` products.
- Adds a new `EarlySwiftDriver` product to that category, which gets built with the host toolchain.
    - Adds an escape hatch: `--skip-early-swift-driver`
- Adjusts the `swift` CMake configuration with an additional step: `swift_create_early_driver_symlinks` which (if one was built) creates symlinks in the `swift` build `bin` directory to the EarlySwiftDriver `swift-driver` and `swift-help` executables.
- Adds a new test subset : `only_early_swiftdriver`, which will get built into a corresponding CMake test target: `check-swift-only_early_swiftdriver-*` which runs a small subset of driver-related tests against the Early SwiftDriver.
    - This subset is run always when the compiler itself is tested (`--test` is specified)
        - With an escape disable-switch: `--skip-test-early-swift-driver` 
    - All tests outside of `only_early_swiftdriver` are forced to run using the legacy C++ driver (to ensure it gets tested, still).
    - NOTE: `SwiftDriver` product (no 'Early') is still the main product used to build the driver for toolchain installation into and for executing the product's own tests. This change does not affect that.

artemcm commented:
@swift-ci please test

artemcm commented:
@nkcsgexi curious what you think of this approach

nkcsgexi commented:
This is a great idea to enable compiler engineers to always test swift-driver as they do for the C++ driver. Thanks for working on this!

artemcm commented:
Together with: https://github.com/apple/swift-driver/pull/537
@swift-ci please smoke test

artemcm commented on utils/swift_build_support/swift_build_support/products/swiftdriver.py:
```diff
@@ -83,8 +87,14 @@ def run_build_script_helper(action, host_target, product, args):
+    if local_compiler_build:
+      # If building with the host toolchain for use with a local compiler build,
+      # use the toolchain which is supplying the `swiftc`.
+      toolchain_path = os.path.dirname(os.path.dirname(os.path.abspath(product.toolchain.swiftc)))
```
It feels like there should be a better way to find this path, but I haven't been able to find one yet.

artemcm commented:
Together with: apple/swift-driver#537
@swift-ci please smoke test Linux platform

artemcm commented:
Together with: apple/swift-driver#537
@swift-ci please smoke test

artemcm commented:
Together with: apple/swift-driver#537
@swift-ci please smoke test macOS Platform

artemcm commented:
Together with: apple/swift-driver#537
@swift-ci please smoke test macOS Platform

artemcm commented:
Together with: apple/swift-driver#537
@swift-ci please smoke test macOS Platform

artemcm commented:
Together with: apple/swift-driver#537
@swift-ci please smoke test macOS Platform

artemcm commented:
Together with: apple/swift-driver#537
@swift-ci please smoke test macOS Platform

artemcm commented:
Together with: apple/swift-driver#537
@swift-ci please smoke test macOS Platform

artemcm commented:
Together with: apple/swift-driver#537
@swift-ci please smoke test macOS Platform

artemcm commented:
Together with: apple/swift-driver#537
@swift-ci please smoke test macOS Platform

artemcm commented:
Together with: apple/swift-driver#537
@swift-ci please smoke test macOS Platform

artemcm commented:
Together with: apple/swift-driver#537
@swift-ci please smoke test macOS Platform

artemcm commented:
Together with: apple/swift-driver#537
@swift-ci please smoke test Linux Platform

artemcm commented:
I'm making progress on tests here, but now hitting some LLDB test failures when using the new driver. Investigating...

artemcm commented:
Together with: apple/swift-driver#537
@swift-ci please smoke test

artemcm commented:
Together with: apple/swift-driver#537
@swift-ci please smoke test macOS Platform

artemcm commented:
The LLDB test failures turned out to be real driver issues so addressing those first, before proceeding with this further.

artemcm commented:
Together with apple/swift-driver#582
@swift-ci please test

artemcm commented:
Together with apple/swift-driver#582
@swift-ci please test

artemcm commented:
Together with apple/swift-driver#582
Together with https://github.com/apple/swift-package-manager/pull/3391
@swift-ci please test

artemcm commented:
Together with apple/swift-driver#582
Together with apple/swift-package-manager#3391
@swift-ci please test macOS platform

artemcm commented:
Together with:
https://github.com/apple/swift-tools-support-core/pull/208
https://github.com/apple/swift-driver/pull/595
@swift-ci please test

gottesmm commented:
One thing that I think is important is that we add documentation on how to invoke the build-script helper here to debuggingthecompiler.md and its relation to dry-run.

A few people have reached out to me and I told this information and it assuaged their concerns. Can you add that? This is information that we may also want to add to build-script's help file.

gottesmm commented on cmake/modules/SwiftUtils.cmake:
```diff
@@ -152,6 +152,43 @@ function(swift_create_post_build_symlink target)
 
+function(swift_create_early_driver_symlinks target)
+  # Is there a better way to get the top-level `build` dir prefix?
+  set(driver_bin_dir "${CMAKE_CURRENT_BINARY_DIR}/../../../earlyswiftdriver-${SWIFT_HOST_VARIANT}-${SWIFT_HOST_VARIANT_ARCH}/release/bin")
```
This should be passed into CMake on the command line.

artemcm replied:
Used CMake's existing specification of the `swift` build directory that was passed into it on the command line instead.

gottesmm commented on cmake/modules/SwiftUtils.cmake:
```diff
@@ -152,6 +152,43 @@ function(swift_create_post_build_symlink target)
+
+  message("From: ${driver_bin_dir}/swift-help")
+  message("To: ${swift_bin_dir}/swift-help")
+  add_custom_command(TARGET "${target}" POST_BUILD
```
I think we have some cmake wrappers for creating symlinks. Can you git grep around for that.

artemcm replied:
Replaced with the existing `swift_create_post_build_symlink`. 👍

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test

artemcm commented:
@gottesmm, comments above addressed and I wrote up some docs on debugging this flow. Could you please take another look?

gottesmm commented on docs/DebuggingTheCompiler.md:
```diff
@@ -807,6 +812,74 @@ reducing SIL test cases by:
 
+## Debugging the Compiler Driver
+
+The Swift compiler uses a standalone compiler-driver application:
```
I would make it clear here that the application is written in swift.

artemcm replied:
Done.

gottesmm commented on docs/DebuggingTheCompiler.md:
```diff
@@ -807,6 +812,74 @@ reducing SIL test cases by:
+## Debugging the Compiler Driver
+
+The Swift compiler uses a standalone compiler-driver application:
+[swift-driver](https://github.com/apple/swift-driver) When building the compiler
```
'.' before When.

artemcm replied:
Done. Thanks.

gottesmm commented on docs/DebuggingTheCompiler.md:
```diff
@@ -807,6 +812,74 @@ reducing SIL test cases by:
+The Swift compiler uses a standalone compiler-driver application:
+[swift-driver](https://github.com/apple/swift-driver) When building the compiler
+using `build-script`, by default, the standalone driver will be built first,
+using the host toolchain. Once the compiler is built, the compiler build
```
I would add "if the host toolchain contains swift. Today, we will not error if the host toolchain does not contain swift, but that may change in the future.'

artemcm replied:
Added clarification. Thanks.

gottesmm commented on cmake/modules/SwiftUtils.cmake:
```diff
@@ -152,6 +152,34 @@ function(swift_create_post_build_symlink target)
+      return()
+  endif()
+
+  message("Creating early SwiftDriver symlinks.")
```
`message(STATUS )`

artemcm replied:
Fixed. Thanks.

gottesmm commented on utils/swift_build_support/swift_build_support/products/earlyswiftdriver.py:
```diff
@@ -0,0 +1,133 @@
+
+# A "Special" SwiftDriver that gets build with the host toolchain to act
+# as *the* compiler driver for the `swift` build directory compiler,
+# hence it does not depend on any other build product of `build-script`.
```
I would add more documentation here around the special behavior of this specifically:

1. I would add that we don't build if we don't find in the host toolchain. But we may error in the future.
2. I would add that the application is written in Swift.
....

The big thing for me is that I want someone who is reading earlyswiftdriver.py/swiftdriver.py to be able to understand the difference in between them without searching through the rest of the repository.

artemcm replied:
Added a longer explanation here, and a comment to `swiftdriver.py`.

gottesmm reviewed (commented):
Some initial comments. I am going to go read the cmake offline.

gottesmm commented:
@atrick can you take a look at the docs in a little bit (after @artemcm says he has updated it) and tell us if it answers any questions you have in terms of being able to still build without build-script.

gottesmm commented on test/CMakeLists.txt:
```diff
@@ -353,6 +354,7 @@ foreach(SDK ${SWIFT_SDKS})
              (test_subset STREQUAL "validation") OR
              (test_subset STREQUAL "only_long") OR
              (test_subset STREQUAL "only_stress") OR
+             (test_subset STREQUAL "only_early_swiftdriver") OR
```
Shouldn't you also need to add only_early_swiftdriver to the if statement on 361?

artemcm replied:
I do not think so. Here, depending on which test subset target is being built, we are adding a list of which test directories need to be passed to `lit`. For the `only_early_swiftdriver` subset, we do need the `test_bin_dir`, because that subset only runs tests inside `test/Driver/`. We do not need to add the `validation_test_bin_dir` for this subset, because it does not require any of the tests from that directory. 

The same build may also be performing validation testing, but that will be built as a different target, with a different subset, which will include test directories relevant to it.

gottesmm replied:
alright. I am fine with this.

gottesmm commented on test/Unit/lit.cfg:
```diff
@@ -63,7 +63,7 @@ elif swift_test_subset == 'only_stress':
     pass
 else:
-    lit_config.fatal("Unknown test mode %r" % swift_test_subset)
+    lit_config.fatal("Unknown test subset %r" % swift_test_subset)
```
What about if we run long tests... why not run long tests with early swift driver?

artemcm replied:
The intent is to have a small set of tests that verify the early driver's basic functionality. 

When running compiler's test suites (such as the long tests), we run them with the legacy driver, because that too needs validation. But most importantly, when building and testing the actual SwiftDriver build product (not early), it has a setup where it will run driver-relevant lit tests https://github.com/apple/swift-driver/blob/main/Tests/SwiftDriverTests/IntegrationTests.swift#L103. This is where we should actually thoroughly test things, and consider running the long tests. 

There are two separate directions for the future:
- Reduce the lit tests that exercise the legacy driver over time to verify only the most basic functionality, transitioning them to instead run when testing the new driver.
- Gradually port over driver-specific tests to the driver repository's XCTests, instead of lit.

gottesmm replied:
Ok. I am fine with this. I was just curious if this was an over sight.

gottesmm commented on cmake/modules/SwiftUtils.cmake:
```diff
@@ -152,6 +152,34 @@ function(swift_create_post_build_symlink target)
+  set(swift_bin_dir "${SWIFT_RUNTIME_OUTPUT_INTDIR}")
+  # If early swift-driver wasn't built, nothing to do here.
+  if(NOT EXISTS "${driver_bin_dir}/swift-driver" OR NOT EXISTS "${driver_bin_dir}/swift-help")
+      return()
```
Can you add a log message here telling the user that we are not creating the symlinks and why?

artemcm replied:
Done:
```
message(STATUS "Skipping creating early SwiftDriver symlinks - no early SwiftDriver build found.")
```

gottesmm replied:
Thanks! (In general I like to add some sort of status in exceptional cases to help debugging misconfiguration).

gottesmm commented on utils/build-script:
```diff
@@ -848,14 +848,21 @@ class BuildScriptInvocation(object):
+        # If --skip-early-swift-driver is passed in, swift will be built
+        # as usual, but relying on its own C++-based (Legacy) driver.
+        # Otherwise, we build an "early" swift-driver using the host
+        # toolchain, which the later-built compiler will forward
```
I would add here that we do this only if we find swiftc in the host toolchain.

artemcm replied:
Clarified. Thanks.

gottesmm commented on utils/build-script:
```diff
@@ -986,10 +997,48 @@ class BuildScriptInvocation(object):
+                    continue
+
+                product_source = product_class.product_source_name()
+                product_name = product_class.product_name()
```
One thing that we may want to do is refactor out a part of this into a helper.

artemcm replied:
Done.

gottesmm commented on utils/build_swift/build_swift/driver_arguments.py:
```diff
@@ -193,6 +193,11 @@ def _apply_default_arguments(args):
 
+    # --test implies --test-early-swift-driver
+    # (unless explicitly skipped with `--skip-test-early-swift-driver`)
+    if args.test and args.test_early_swift_driver is None:
```
What happens here if we don't build swift driver? Please add tests for the failure modes here.

artemcm replied:
Changed to only trigger `test_early_swift_driver` if early driver is being built.
If we expect early driver to get built, but the host toolchain did not have a Swift compiler, this will still run the driver test subset, it'll just run against the legacy driver.

gottesmm replied:
Seems like something we would want to warn/error to provide visibility. Consider this running on a bot and we just silently stopped testing early swift driver and were testing the legacy driver the whole time. That would be unfortunate. Can you add a warning in such a scenario at minimum?

gottesmm commented on utils/swift_build_support/swift_build_support/host_specific_configuration.py:
```diff
@@ -167,6 +167,14 @@ def __init__(self, host_target, args):
+                # If the compiler is being tested after being built to use the
+                # standalone swift-driver, we build a test-target to
+                # run a reduced set of lit-tests that verify the early swift-driver.
+                if getattr(args, 'test_early_swift_driver', False) and\
```
I am pretty sure that the getattr here is not necessary. But I am ok with it.

gottesmm commented on utils/swift_build_support/swift_build_support/products/earlyswiftdriver.py:
```diff
@@ -0,0 +1,133 @@
+        '--local_compiler_build'
+    ]
+
+    if os.path.exists(dispatch_build_dir):
```
Question. Should we error here if we do not have these build directories? How does the failure of these conditions manifest?

artemcm replied:
As far as I remember, this is platform specific. On Linux, `build-script` propagates an explicit Foundation build dir to the products, which is required. If that is missing, then the driver's `build-script-helper.py` will fail. On Darwin, it doesn't matter because we expect Foundation to just be found on the host system. 

In any case, I think if there is an issue with this, we should fix it separately from this change, for both the (driver) build products.

gottesmm replied:
Ok. SGTM. I was just thinking about the failure modes on Linux.

gottesmm commented on utils/swift_build_support/swift_build_support/products/product.py:
```diff
@@ -58,6 +58,14 @@ def is_build_script_impl_product(cls):
+    def is_before_build_script_impl_product(cls):
+        """is_before_build_script_impl_product -> bool
+
+        Whether this product is build before any build-script-impl products.
```
I would add to the doc here that there is an assumption that this is a non-build-script-impl product.

Also, it may be worth it to use a NotImplementedError and then to override it in sub-classes if there is a nice place to do it. That way if someone forgets they get an error. That being said, I am not completely wedded to that.

artemcm replied:
Done, and added `NotImplementedError` here.

gottesmm replied:
Looks good!

artemcm commented:
@atrick The docs should be in a good shape now for you to take a look. Thanks!

atrick reviewed (commented):
Thanks for adding all the build-script info to the docs! Looks good to me.

I'm mostly curious about how to understand/debug differences between the old/new driver and other driver-like tools.

Will 'swiftc ... -###' always print the same set of commands for the old/new driver? Do they call 'swift-frontend' the same way?

What's the difference between invoking 'swiftc' vs. 'swift-driver' at the top level?

What's the difference between invoking 'swift -frontend' vs. 'swift-frontend' (I sure hope its nothing)?

If there's a 'swift -frontend' to invoke 'swift-frontend' why isn't there a 'swift -c' to invoke 'swift-driver'?

These don't need to be answered in this PR, but I didn't happen to see an explanation in the docs and these are the important questions to me.

gottesmm commented on docs/DebuggingTheCompiler.md:
```diff
@@ -807,6 +814,89 @@ reducing SIL test cases by:
+A "dry-run" invocation of the `build-script` (using the `--dry-run` flag) will
+print the commands that would be executed in a given build, without executing
+them. A dry-run script invocation output can be used to inspect the build stages
+of a given `build-script` configuration, or create script corresponding to one
```
Beautiful!

gottesmm reviewed (commented):
Looks reasonable to me. My suggestion is why not just answer Andy's questions directly in an FAQ section on the swift-driver. Just one more section in the docs. I am fine if you want to do that in a subsequent PR so we can get this in. If you want to get it in this PR as well, your choice.

gottesmm commented:
The nice thing about the FAQ is that rather than having to deal with narrative, you can just say... some answers to common questions or something like that so you can write it quickly. If you can find a way to fit it into the narrative feel free. Just don't want to hold up this work further.

atrick commented:
I'm not sure if my questions even make sense, so you don't need to repeat them verbatim in a doc. I'm confused about the design of the command line interfaces for our driver-like tools and not sure where to look.

artemcm commented:
> Thanks for adding all the build-script info to the docs! Looks good to me.
> 
> I'm mostly curious about how to understand/debug differences between the old/new driver and other driver-like tools.
> 
> Will 'swiftc ... -###' always print the same set of commands for the old/new driver? Do they call 'swift-frontend' the same way?

The new swift-driver (Still invoked as `swiftc`) is meant to b a direct drop-in replacement. As such it has the exact same command-line interface. The expectation is that its behaviour closely matches the legacy driver; however, during, and after the transition to the new driver being the default its behaviour may start to diverge from the legacy driver as par for the course of its evolution and gaining new features, etc. Today, broadly-speaking, `swift-frontend` invocations generated by the two drivers are expected to be very similar. 

> 
> What's the difference between invoking 'swiftc' vs. 'swift-driver' at the top level?

The mechanism by which we are deploying the new driver today is as follows:
- The user invokes `swiftc`, which is the same executable as before, which forwards the invocation to the `swift-driver` executable if one is found alongside it. There are two escape hatches that will make `swiftc` avoid forwarding to the new driver:
- `-disallow-use-new-driver` command line flag
- `SWIFT_USE_OLD_DRIVER` environment variable
These are meant to allow for a fallback to the legacy driver during the transition. 
If the user is to directly invoke the `swift-driver` executable, the behaviour should be the same as invoking `swiftc` executable, but without the option for a legacy driver fallback. 

> 
> What's the difference between invoking 'swift -frontend' vs. 'swift-frontend' (I sure hope its nothing)?

AFAIK, there is no difference in doing this with either driver. 

> 
> If there's a 'swift -frontend' to invoke 'swift-frontend' why isn't there a 'swift -c' to invoke 'swift-driver'?

`swift -frontend` bypasses any and all driver logic and invokes the compiler directly. With the new driver, once we go through this transition, `swift` and `swiftc` executables themselves will be the new driver executables (symlinks to `swift-driver`). So it seems that `swift -c` is a bit redundant. Unless I misunderstood this part? 

> 
> These don't need to be answered in this PR, but I didn't happen to see an explanation in the docs and these are the important questions to me.

@atrick, thank you for these questions, I do think they highlight some pretty obvious gaps in documentation in this transition which must be documented, either in the docs here, or in the swift-driver repo, or both. 

I like @gottesmm 's idea of having an F.A.Q. in the docs here, with the swift-driver's own documentation (in its corresponding repository) providing more in-depth docs.

artemcm commented:
@swift-ci please test

artemcm committed:
Always build (and use) the new SwiftDriver as the default compiler driver.

This will make sure that compiler developers are using the new driver when they build the compiler locally and use it.

- Adds a new build-script product category: before_build_script_impl for products we wish to build before the impl products.
- Adds a new EarlySwiftDriver product to that category, which gets built with the host toolchain.
- Adds an escape hatch: --skip-early-swift-driver
- Adjusts the swift CMake configuration with an additional step: swift_create_early_driver_symlinks which (if one was built) creates symlinks in the swift build bin directory to the EarlySwiftDriver swift-driver and swift-help executables.
- Adds a new test subset : only_early_swiftdriver, which will get built into a corresponding CMake test target: check-swift-only_early_swiftdriver-* which runs a small subset of driver-related tests against the Early SwiftDriver.
  - This subset is run always when the compiler itself is tested (--test is specified)
  - With an escape disable-switch: --skip-test-early-swift-driver
  - All tests outside of only_early_swiftdriver are forced to run using the legacy C++ driver (to ensure it gets tested, still).

NOTE: SwiftDriver product (no 'Early') is still the main product used to build the driver for toolchain installation into and for executing the product's own tests. This change does not affect that.

artemcm committed:
Add documentation about the compiler's standalone driver and its build procedure.

artemcm commented:
@swift-ci please test

atrick commented:
Thanks Artem!

I think you can just cross reference `Driver.md` from this "debugging" document. Explaining the command line interface probably goes better in the driver docs.

> swift -frontend bypasses any and all driver logic and invokes the compiler directly. With the new driver, once we go through this transition, swift and swiftc executables themselves will be the new driver executables (symlinks to swift-driver). So it seems that swift -c is a bit redundant. Unless I misunderstood this part?

`swift -c` is (I think) redundant with `swiftc`. My point is that `swift -frontend` is also (I think) redundant with `swift-frontend`. Is there a reason for the redundancy in one case and not the other. I don't have an opinion, I'm just trying to understand the CLI.

Also, `swift` has a different CLI from `swiftc`. So which one is the driver interface?

artemcm commented:
>swift -c is (I think) redundant with swiftc. My point is that swift -frontend is also (I think) redundant with swift-frontend. Is there a reason for the redundancy in one case and not the other. I don't have an opinion, I'm just trying to understand the CLI.

That's a good point, it is redundant, and is just an artifact of the times before `swift-frontend`. The only thing keeping it around is that we may still have compiler clients who may be relying on it, e.g. things like SwiftPM may still be generating `swift -frontend` invocations. We should be able to deprecate it if we make sure the clients know to invoke `swift-frontend` directly.  

>Also, swift has a different CLI from swiftc. So which one is the driver interface?

Both are the driver, just two different "modes" of the driver. For `swift`, the driver's job is simply to [create](https://github.com/apple/swift-driver/blob/main/Sources/SwiftDriver/Jobs/Planning.swift#L638) the `swift-frontend -interpret` job and setup its arguments correctly. For `swiftc`, he driver's job is to [plan the build](https://github.com/apple/swift-driver/blob/main/Sources/SwiftDriver/Jobs/Planning.swift#L80), as normal. The driver, internally, distinguishes them by knowing whether or not it was invoked as `swift` or `swiftc`. (or `--driver-mode=[swift|swiftc]` if invoking the `swift-driver` directly).

artemcm commented:
@swift-ci please test Windows platform
